mount/mount_id_mapped.go

```diff
@@ -0,0 +1,130 @@
+	}
+	hID, err := strconv.ParseUint(parts[1], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid host id for user namespace remapping")
+	}
+	size, err := strconv.ParseUint(parts[2], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid size for user namespace remapping")
+	}
+
+	return []syscall.SysProcIDMap{
+		{
+			ContainerID: int(cID),
+			HostID:      int(hID),
+			Size:        int(size),
+		},
+	}, nil
+}
+
+func mountIDMapped(target string, pid int) (err error) {
+	var (
+		path       string
+		attr       Attr
+		userNsFile *os.File
+		targetDir  *os.File
+	)
+
+	path = fmt.Sprintf("/proc/%d/ns/user", pid)
+	if userNsFile, err = os.Open(path); err != nil {
+		return errors.Wrapf(err, "Unable to get user ns file descriptor for - %s", path)
+	}
+
+	attr.attrSet = MOUNT_ATTR_IDMAP
+	attr.attrClr = 0
+	attr.propagation = 0
+	attr.userNs = uint64(userNsFile.Fd())
+
+	defer userNsFile.Close()
+	if targetDir, err = os.Open(target); err != nil {
+		return errors.Wrapf(err, "Unable to get mount point target file descriptor - %s", target)
+	}
+
+	defer targetDir.Close()
+	return SetAttr(int(targetDir.Fd()), "", unix.AT_EMPTY_PATH|AT_RECURSIVE,
+		&attr, uint(unsafe.Sizeof(attr)))
+}
+
+// MapMount applies GID/UID shift according to gidmap/uidmap for target path
+func MapMount(uidmap string, gidmap string, target string) (err error) {
+	const (
+		userNsHelperBinary = "/bin/true"
+	)
+	// TODO: Avoid dependency on /bin/true or do in a completely different way
+	// Currently there is no way to pass idmapping directly to mount_setattr,
+	// this is not very convinient from the container runtime point of view.
+	// The id remapping procedure should be done in containerd, due to we have
+	// old approach the use recurcive chown.
+	// Maybe it is necessary to think about moving of container rootfs ownership
+	// adjustment to runc due to runc have knowledge about container user namespace.
+	// But personally I think that it would be better to add possibility to call
+	// mount_setattr with explicit id mappings and leave container runtime components
+	// responsibilities unchanged.
+	cmd := exec.Command(userNsHelperBinary)
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUSER,
+	}
+
+	if cmd.SysProcAttr.UidMappings, err = parseIDMapping(uidmap); err != nil {
+		return err
+	}
+	if cmd.SysProcAttr.GidMappings, err = parseIDMapping(gidmap); err != nil {
+		return err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return errors.Wrapf(err, "Failed to run the %s helper binary", userNsHelperBinary)
+	}
+
+	if err = mountIDMapped(target, cmd.Process.Pid); err != nil {
+		return errors.Wrapf(err, "Failed to create idmapped mount for target - %s", target)
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return errors.Wrapf(err, "Failed to run the %s helper binary", userNsHelperBinary)
+	}
```

Comment:
I think this should be called using defer. Otherwise it won't be called when `mountIDMapped` fails.